hail_v02_scripts/utils/elasticsearch/client.py

```diff
@@ -0,0 +1,253 @@
+                elasticsearch_write_operation="update", and the desired behavior is to update all documents that exist,
+                but to ignore errors for documents that don't exist.
+            elasticsearch_mapping_id (str): if specified, sets the es.mapping.id which is the column name to use as the document ID
+                See https://www.elastic.co/guide/en/elasticsearch/hadoop/current/configuration.html#cfg-mapping
+            field_name_to_elasticsearch_type_map (dict): (optional) a map of keytable field names to
+                their elasticsearch field spec - for example: {
+                    'allele_freq': { 'type': 'half_float' },
+                    ...
+                }.
+                See https://www.elastic.co/guide/en/elasticsearch/reference/current/mapping.html for
+                more details. Any values in this dictionary will override
+                the default type mapping derived from the hail keytable column type.
+                Field names can be regular expressions.
+            disable_doc_values_for_fields (tuple): (optional) list of field names (the way they will be
+                named in the elasticsearch index) for which to not store doc_values
+                (see https://www.elastic.co/guide/en/elasticsearch/reference/current/mapping-params.html)
+            disable_index_for_fields (tuple): (optional) list of field names (the way they will be
+                named in the elasticsearch index) that shouldn't be indexed
+                (see https://www.elastic.co/guide/en/elasticsearch/reference/current/mapping-params.html)
+            field_names_replace_dot_with (string): since "." chars in field names are interpreted in
+                special ways by elasticsearch, set this arg to first go through and replace "." with
+                this string in all field names. This replacement is not reversible (or atleast not
+                unambiguously in the general case) Set this to None to disable replacement, and fall back
+                on an encoding that's uglier, but reversible (eg. "." will be converted to "_$dot$_")
+            export_globals_to_index_meta (bool): whether to add table.globals object to the index _meta field:
+                (see https://www.elastic.co/guide/en/elasticsearch/reference/current/mapping-meta-field.html)
+            verbose (bool): whether to print schema and stats
+        """
+
+        # see https://www.elastic.co/guide/en/elasticsearch/hadoop/master/spark.html#spark-sql-write
+        # "By default, elasticsearch-hadoop will ignore null values in favor of not writing any field at all.
+        # Since a DataFrame is meant to be treated as structured tabular data, you can enable writing nulls as null
+        # valued fields for DataFrame Objects only by toggling the es.spark.dataframe.write.null setting to true."
+        elasticsearch_config = {"es.spark.dataframe.write.null": "true"}
```

Comment:
Not sure if this should be on by default, configurable, or we should just ignore it. I added it here to be consistent with the Hail 0.1 scripts.

https://www.elastic.co/guide/en/elasticsearch/hadoop/master/spark.html#spark-sql-write
> By default, elasticsearch-hadoop will ignore null values in favor of not writing any field at all. Since a DataFrame is meant to be treated as structured tabular data, you can enable writing nulls as null valued fields for DataFrame Objects only by toggling the es.spark.dataframe.write.null setting to true.

For example, for `{foo: 'value', bar: null}`, Hail 0.1 would write `{foo: 'value', bar: null}` to Elasticsearch. Without this configuration setting, Hail 0.2 writes `{foo: 'value'}`.

Reply:
I know it wouldn't delete the fields from the mapping (since ES only lets you add fields to an existing index, but not remove). My guess is it wouldn't touch the values either.